qiskit_experiments/framework/backend_timing_mixin.py

```diff
@@ -0,0 +1,290 @@
+
+        raise QiskitError("Backend has no dt value.")
+
+    def circuit_delay(self, time: float) -> Union[int, float]:
+        """Delay duration close to ``time`` and consistent with timing constraints
+
+        This method produces the value to pass for the ``duration`` of a
+        ``Delay`` instruction of a ``QuantumCircuit`` or a pulse schedule so
+        that the delay fills the time until the next valid pulse, assuming the
+        ``Delay`` instruction begins on a sample that is also valid for a pulse
+        to begin on.
+
+        The pulse timing constraints of the backend are considered in order to
+        give a number of samples closest to ``time`` plus however many more
+        samples are needed to get to the next valid sample for the start of a
+        pulse in a subsequent instruction. The least common multiple of the
+        pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        If :meth:`.BackendTiming.delay_unit` is ``s``, ``time`` is
+        returned directly. Typically, this is the case for a simulator where
+        converting to sample number is not needed.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples if :meth:`.BackendTiming.delay_unit`
+            is ``dt``. Other return ``time``.
+        """
+        if self.delay_unit == "s":
+            return time
+
+        return self.schedule_delay(time)
+
+    def schedule_delay(self, time: float) -> int
+        pulse_alignment = self._backend_data.pulse_alignment
+        acquire_alignment = self._backend_data.acquire_alignment
+
+        granularity = lcm(pulse_alignment, acquire_alignment)
+
+        samples = int(round(time / self._dt / granularity) * granularity)
+
+        return samples
+
+    def pulse_samples(self, time: float) -> int:
+        """The number of samples giving a valid pulse duration closest to ``time``
+
+        Args:
+            time: Pulse duration in seconds
+
+        Returns:
+            The number of samples corresponding to ``time``
+
+        Raises:
+            QiskitError: If the algorithm used to calculate the pulse length
+                produces a length that is not commensurate with the pulse or
+                acquire alignment values. This should not happen unless the
+                alignment constraints provided by the backend do not fit the
+                assumptions that the algorithm makes.
+        """
+        granularity = self._backend_data.granularity
+        min_length = self._backend_data.min_length
+
+        samples = int(round(time / self._dt / granularity)) * granularity
+        samples = max(samples, min_length)
+
+        pulse_alignment = self._backend_data.pulse_alignment
+        acquire_alignment = self._backend_data.acquire_alignment
+
+        if samples % pulse_alignment != 0:
```

Comment:
Any reason to validate alignments here? We can still compensate for it with extra delay. Perhaps this makes sense because we don't always want to invoke the compute-heavy constrained reschedule in transpile.

Reply:
This paranoia on my part. The motivation given in the docstrings says that if one uses the methods here for all pulses and delays that all pulses should have valid alignment. For that to be true, these conditions need to hold. These conditions should always be true for current IBM backends any way. I am just not sure what could possibly happen in the future. It seems likely that the granularity would always be a multiple of the alignment, but I wasn't sure. One could use Waveform to pad with a delay but that might undo the benefit of a shorter pulse in some  cases.

I don't think these checks add much overhead but we could drop them as likely always redundant. What do you think?

Reply:
I think this is reasonable assumption and now it looks good with new documentation. I was concerning that the developer could misunderstand the meaning of these constraints by reading this code, but this should be okey as long as the logic behind the scene is properly described like now.